tests/test_apply.py

```diff
@@ -236,49 +241,209 @@ def test_apply_onto_other_ref(data_working_copy, cli_runner):
-                        }
-                    }
-                },
-            },
-            "kart.patch/v1": {
-                "authorEmail": "robert@example.com",
-                "authorName": "Robert Coup",
-                "authorTime": "2019-06-20T14:28:33Z",
-                "authorTimeOffset": "+01:00",
-                "message": "Change the title",
-            },
+            "meta": {
+                "title": {
+                    "+": "new title:",
+                }
+            }
         }
     )
+    patch["kart.patch/v1"].pop("originalParent")
     with data_archive("points"):
-        # We ordinarily can't apply this patch,
+        # We can't apply this patch,
         # because the "-" object for the title is missing.
-        r = cli_runner.invoke(
-            ["apply", "-"],
-            input=patch_file,
+        with write_patch(patch, tmp_path) as patch_path:
+            r = cli_runner.invoke(["apply", patch_path])
+        assert r.exit_code == PATCH_DOES_NOT_APPLY, r.stderr
+
+
+def test_apply_minimal_style_meta_patch(data_archive, cli_runner, tmp_path):
+    patch = points_patch(
+        {
+            "meta": {
+                "title": {
+                    "+": "new title:",
+                }
+            }
+        }
+    )
+    with data_archive("points"):
+        with write_patch(patch, tmp_path) as patch_path:
+            r = cli_runner.invoke(["apply", patch_path])
+        assert r.exit_code == 0, r.stderr
+
+        # Check that the change was actually applied
+        r = cli_runner.invoke(["show", "-o", "json", "HEAD"])
+        assert r.exit_code == 0
+        show = json.loads(r.stdout)
+        meta = show["kart.diff/v1+hexwkb"]["nz_pa_points_topo_150k"]["meta"]
+        assert meta == {"title": {"+": "new title:", "-": "NZ Pa Points (Topo, 1:50k)"}}
+
+        # now check that a conflict gets rejected
+        # note: we change the title here but don't change the originalParent,
+        # so this change now conflicts with the change we just applied.
+        patch = points_patch(
+            {
+                "meta": {
+                    "title": {
+                        "+": "differently new title:",
+                    }
+                }
+            }
         )
+        with write_patch(patch, tmp_path) as patch_path:
+            r = cli_runner.invoke(["apply", patch_path])
         assert r.exit_code == PATCH_DOES_NOT_APPLY, r.stderr
 
-        # But --allow-missing-old-values lets us do it.
-        r = cli_runner.invoke(
-            ["apply", "--allow-missing-old-values", "-"],
-            input=patch_file,
+
+def test_apply_minimal_style_feature_patch_with_edit(
+    data_archive, cli_runner, tmp_path
+):
+
+    patch = points_patch(
+        {
+            "feature": [
+                {
+                    # one edit (in minimal style it looks like an insert, but FID 1182 already exists)
+                    "+": {
+                        "fid": 1182,
```

Comment:
I haven't reviewed it all, but this is reasonably fundamental (maybe should be on #477): should the implementation of this be passing the commit as a parameter rather than putting it _in_ the payload?

The comment here worries me a bit that we're writing patches that look like inserts but are actually edits.

Reply:
> should the implementation of this be passing the commit as a parameter rather than putting it in the payload?

I'm unsure what you're suggesting here. The original parent commit hash is stored in the patch header (`originalParent`, feel free to suggest a better name; I don't really like that key but I couldn't think of a better one) because it's part of the patch - omitting it would mean we can't apply this patch because we don't know the starting state and hence can't resolve conflicts when the `-` blobs are missing.


> The comment here worries me a bit that we're writing patches that look like inserts but are actually edits.

what if we changed the key from `+` to something else for updates in minimal patches? e.g. `-+`? I'd be happy to do that, it would make the intent of the patch much clearer.

Reply:
> I'm unsure what you're suggesting here.

Would something like `kart apply <TO-WHAT> <PATCH-FILE>` → `kart apply c0ffee my.patch` be clearer or more/less useful? Compared with putting `c0ffee` _into_ `my.patch`

Reply:
I don't think so - without that `c0ffee` the patch itself is useless and can't be applied, so it doesn't make sense to me to exclude that info from the patch. It isn't the thing you're applying the patch _to_ (you're always applying the patch to HEAD) but it tells kart what state you generated the patch _from_. You're right that we _could_ pass this info in a sideband but I don't see why we would - it's an indispensible part of the patch. Without that all we have is the state of a bunch of features from some unknown source...

Reply:
put it another way: at `apply` time i wouldn't expect the user to know what revision the patch was generated from. So they shouldn't have to specify it on the command line

Reply:
Makes sense, thanks 👍

Reply:
The second issue:

>> The comment here worries me a bit that we're writing patches that look like inserts but are actually edits.
>
> what if we changed the key from "+" to something else for updates in minimal patches? e.g. "-+"?

I'm keen on something instead of `+`, and I'm not excited about `'-+'` (or `'+-'`). I'd probably prefer something from `*$~!@` etc